Bottom: 0b16364efabd9cd0ac814816e5b89103e70b2739
Top:    aa28be9cf5344669994716a5f84b8a0f7e4c8455
Author: Roland Stigge <stigge@antcom.de>
Date:   2012-03-29 23:31:15 +0200

Fix i2c start on PNX

This patch fixes i2c start by checking for the right mif.mode

Signed-off-by: Roland Stigge <stigge@antcom.de>


---

diff --git a/drivers/i2c/busses/i2c-pnx.c b/drivers/i2c/busses/i2c-pnx.c
index e0fea61..83ae3b3 100644
--- a/drivers/i2c/busses/i2c-pnx.c
+++ b/drivers/i2c/busses/i2c-pnx.c
@@ -137,21 +137,23 @@ static int i2c_pnx_start(unsigned char slave_addr,
 		return -EINVAL;
 	}
 
-	/* First, make sure bus is idle */
-	if (wait_timeout(alg_data)) {
-		/* Somebody else is monopolizing the bus */
-		dev_err(&alg_data->adapter.dev,
-			"%s: Bus busy. Slave addr = %02x, cntrl = %x, stat = %x\n",
-			alg_data->adapter.name, slave_addr,
-			ioread32(I2C_REG_CTL(alg_data)),
-			ioread32(I2C_REG_STS(alg_data)));
-		return -EBUSY;
-	} else if (ioread32(I2C_REG_STS(alg_data)) & mstatus_afi) {
-		/* Sorry, we lost the bus */
-		dev_err(&alg_data->adapter.dev,
-		        "%s: Arbitration failure. Slave addr = %02x\n",
-			alg_data->adapter.name, slave_addr);
-		return -EIO;
+	if (alg_data->mif.mode == 0) {
+		/* First, make sure bus is idle */
+		if (wait_timeout(alg_data)) {
+			/* Somebody else is monopolizing the bus */
+			dev_err(&alg_data->adapter.dev,
+				"%s: Bus busy. Slave addr = %02x, cntrl = %x, stat = %x\n",
+				alg_data->adapter.name, slave_addr,
+				ioread32(I2C_REG_CTL(alg_data)),
+				ioread32(I2C_REG_STS(alg_data)));
+			return -EBUSY;
+		} else if (ioread32(I2C_REG_STS(alg_data)) & mstatus_afi) {
+			/* Sorry, we lost the bus */
+			dev_err(&alg_data->adapter.dev,
+				"%s: Arbitration failure. Slave addr = %02x\n",
+				alg_data->adapter.name, slave_addr);
+			return -EIO;
+		}
 	}
 
 	/*
